debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/processor/ehcache/EhcacheLogMinerTransactionCache.java

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.connector.oracle.logminer.processor.ehcache;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import org.ehcache.Cache;
+
+import io.debezium.connector.oracle.logminer.events.LogMinerEvent;
+import io.debezium.connector.oracle.logminer.processor.AbstractLogMinerTransactionCache;
+
+/**
+ * A concrete implementation of {@link AbstractLogMinerTransactionCache} for Ehcache.
+ *
+ * @author Chris Cranford
+ */
+public class EhcacheLogMinerTransactionCache extends AbstractLogMinerTransactionCache<EhcacheTransaction> {
+
+    private final Cache<String, EhcacheTransaction> transactionCache;
+    private final Cache<String, LogMinerEvent> eventCache;
+
+    // Heap-backed caches for quick access to specific metadata to speed up processing
+    private final Map<String, TreeSet<Integer>> eventIdsByTransactionId = new HashMap<>();
+
+    public EhcacheLogMinerTransactionCache(Cache<String, EhcacheTransaction> transactionCache, Cache<String, LogMinerEvent> eventCache) {
+        this.transactionCache = transactionCache;
+        this.eventCache = eventCache;
+    }
+
+    @Override
+    public EhcacheTransaction getTransaction(String transactionId) {
+        return transactionCache.get(transactionId);
+    }
+
+    @Override
+    public void addTransaction(EhcacheTransaction transaction) {
+        transactionCache.put(transaction.getTransactionId(), transaction);
+        eventIdsByTransactionId.put(transaction.getTransactionId(), new TreeSet<>());
+    }
+
+    @Override
+    public void removeTransaction(EhcacheTransaction transaction) {
+        transactionCache.remove(transaction.getTransactionId());
```

Comment:
Should eventIdsByTransactionId.remove(transaction.getTransactionId()) be called here?

Reply:
So I can change this to do that, but doing so means more changes in the _processor_ class.  There we actually do the event removal and the transaction removal in two steps.  I'll take a look and see that changes fall out from doing this.